namespace-lock.go

```diff
@@ -0,0 +1,148 @@
+/*
+ * Minio Cloud Storage, (C) 2016 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package main
+
+import "sync"
+
+// nsParam - carries name space resource.
+type nsParam struct {
+	volume string
+	path   string
+}
+
+// nsLock - provides primitives for locking critical namespace regions.
+type nsLock struct {
+	rwMutex *sync.RWMutex
+	count   uint
+}
+
+func (ns *nsLock) InUse() bool {
+	return ns.count != 0
+}
+
+// Lock acquires write lock and increments the namespace counter.
+func (ns *nsLock) Lock() {
+	ns.rwMutex.Lock()
+	ns.count++
+}
+
+// Unlock releases write lock and decrements the namespace counter.
+func (ns *nsLock) Unlock() {
+	ns.rwMutex.Unlock()
+	if ns.count != 0 {
+		ns.count--
+	}
+}
+
+// RLock acquires read lock and increments the namespace counter.
+func (ns *nsLock) RLock() {
+	ns.rwMutex.RLock()
+	ns.count++
+}
+
+// RUnlock release read lock and decrements the namespace counter.
+func (ns *nsLock) RUnlock() {
+	ns.rwMutex.RUnlock()
+	if ns.count != 0 {
+		ns.count--
+	}
+}
+
+// newNSLock - provides a new instance of namespace locking primitives.
+func newNSLock() *nsLock {
+	return &nsLock{
+		rwMutex: &sync.RWMutex{},
+		count:   0,
+	}
+}
+
+// nsLockMap - namespace lock map, provides primitives to Lock,
+// Unlock, RLock and RUnlock.
+type nsLockMap struct {
+	lockMap      map[nsParam]*nsLock
```

Comment:
nsMutexMap map[nsParm]*sync.RWMutex
ref int
Notes: count becomes ref. nsLock object is not necessary. Attach those methods to nsLockMap directly. It is not reusable elsewhere.